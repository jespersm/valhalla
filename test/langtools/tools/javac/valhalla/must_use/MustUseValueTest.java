/*
 * @test /nodynamiccopyright/
 * @summary Xxx.
 * @enablePreview
 *
 * @compile/fail/ref=MustUseValueTest.out -XDrawDiagnostics MustUseValueTest.java
 */
public value final class MustUseValueTest implements java.io.Serializable {

    public static void main(String[] args) {
        var array = new String[] { "A", "B", "C",  };

        Cursor c = new Cursor(array, 0, 3);
        c.next(); // bad code - ERROR
        c = c.next(); // good code - NOT ERROR

        System.out.println("Should be B: " + c.current());
    }

    value record Cursor<T>(T[] array, int position, int length) {
        Cursor {
            if (position < 0 || position >= length) {
                throw new IndexOutOfBoundsException("index %d outside array bounds [0;%d[".formatted(position, length));
            }
        }

        T current() {
            return array[position];
        }

        @MustUse
        Cursor<T> next() {
            return new Cursor(array, position+1, length);
        }
    }
}
