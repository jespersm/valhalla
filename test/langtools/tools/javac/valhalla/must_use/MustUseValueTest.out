MustUseValueTest.java:14:10: compiler.err.method.returned.must.use.value: MustUseValueTest.Cursor, next()
- compiler.note.unchecked.filename: MustUseValueTest.java
- compiler.note.unchecked.recompile
- compiler.note.preview.filename: MustUseValueTest.java, DEFAULT
- compiler.note.preview.recompile
1 error
